sig-cli/charter.md

```diff
@@ -5,25 +5,81 @@ the Roles and Organization Management outlined in [sig-governance].
 
 ## Scope
 
-The Command Line Interface SIG (SIG CLI) is responsible for kubectl and
-related tools. This group focuses on general purpose command line tools and
-libraries to interface with Kubernetes API's.
-
 ### In scope
 
-SIG CLI [README]
-
-#### Code, Binaries and Services
-
-SIG CLI code include general purpose command line tools and binaries for working
-with Kubernetes API's. Examples of these binaries include: [kubectl and kustomize].
+The Command Line Interface SIG (SIG CLI) is responsible for kubectl and
+related tools that are developed by the SIG.  See kubectl
+[design principles](design-principles.md) for the focus of kubectl functionality.
+
+This group focuses on command line tooling for working with Kubernetes APIs and Resource
+Config.  This includes both generalized tooling for working with Resources, Resource Config
```

Comment:
What does working with Resource Config mean? The configuration as stored in Kubernetes and accessible via the API, configuration files as they are stored and worked on outside of the API (e.g., on local disk), or both?

Reply:
BTW, the "on local disk" reference is within scope. kubectl has always read from local disk and, optionally, dumps output of operations, currently to stdout, but I wouldn't preclude writing to file if users want that. 

kubectl and the API were co-designed exactly to support this kind of usage, without necessarily requiring (but not precluding) a different representation, in order to avoid the problem of the API becoming a purely internal interface, hidden beneath some other layer, as well as avoiding compounding other challenges, such as versioning of resource types. This is deeply engrained in the design of Kubernetes. That's described in the resource-model doc, the [architecture doc](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/architecture.md), and other docs.

Reply:
That makes this an incredibly broad scope that crosses over quite a bit with the ecosystem. For example, there are already numerous tools to do resource validation of assets on local disk, in the ecosystem. And, this is just one area.

Should the scope be this broad and overlapping with tools in the ecosystem? What's the benefit for k8s to invest in this?

Reply:
Because end users need to apply minimal config every where a kube distribution happens.  And minimal = takes kube objects and puts them on a server correctly.  Declarative config for kube is well within scope.  Templating and transformation is not.  

I agree with Brian.

Reply:
Just to be clear: I think that the 'kustomize' work falls under "Templating and complex transformation/expansion" do you agree with that @smarterclayton ?

Reply:
The core thing that kustomize does is apply patches, which has been supported by kubectl and by the API for more than 4 years. It does not support variable substitution, it doesn't support conditionals or comprehensions, it's not a configuration language like jsonnet or hcl or starlark, it's not a client library like Pulumi's Typescript library, and it doesn't use an alternative schema like kompose. It uses the raw API schema, in vanilla yaml or json, and strategic merge patch. Any tool that understands API patches can understand its patches. It adds declarative support for existing imperative operations, which has long been requested by users (list of issues are in the KEP). It provides a declarative, bulk version of long-standing singleton imperative operations. It adheres to both the letter and spirit of the project's long-standing documented scope. It is intended to be composable with other approaches and tools, and there are already examples of it being used together with other existing tools.

Reply:
This is where I get a bowl of popcorn ready :)

I don't think we're going to agree on a set of specific statements about what makes something "complex" or not, or when a yaml-encoded template is a "transformation" or "just a patch".  Particularly when we already have a set of specific projects in mind and we're trying to retrofit our preferences onto it as well.

So now how do we agree to disagree?  Afaics we can either let more things in scope, push more things out of scope, or cherry-pick specific projects (and not pretend otherwise).

I think the implied statement with sig-cli is that tools _in_ sig-cli are strongly endorsed as part of the "standard k8s experience", whereas tools _outside_ sig-cli (eg: helm) are optional, and may vary in approach, quality, etc.  Since we want the "standard k8s experience" to be universal, this implies it should be easy (or at least plausible) to re-implement it across multiple (non-golang) languages.  So one rule-of-thumb sig-cli criteria might be "re-implementation difficulty".  This encourages staying close to the API (good), and reusing widely-implemented standards (eg: json patch; good).  We will have to avoid just-because-its-trivial mis-features like envsubst templating, but I think we can navigate around those hazards.

Just to talk about the elephant in the room: kustomize is something like 20k lines of go, which I would not attempt to re-implement in multiple languages, and thus I would consider "not suitable as standard k8s" under the above criteria.  A simpler alternative that was only a stack of json-patches (or even client-side schema-driven s-m-p's) might pass, however.

Reply:
> complex transformation

I don't consider kustomize complex transformation, just patch application with conventions.  It's the simplest possible thing that addresses having a declarative state with a bounded number of variants.  I worked pretty closely with sig-cli to craft a set of boundaries that it shouldn't cross that keep it focused on "ensure local declarative state converges to remote state" without having unbounded scope.  Templatization, language constructs, etc. Linear patch application is annoying to script, but common in practice.

> Just to talk about the elephant in the room: kustomize is something like 20k lines of go, which I would not attempt to re-implement in multiple languages, and thus I would consider "not suitable as standard k8s" under the above criteria

I don't know if I'm missing something from your argument, but I don't generally consider lines of code and reimplementation difficulty as a suitability metric.  kubectl has single commands that are easily 5k of go that could be reimplemented in 5 lines of python, but the extra 4995 lines are there to prevent someone from having to worry about writing the other 4995 lines of python .

Reply:
One useful thing about reference implementations is that they expose when things are more complicated than they should be.

Reply:
FWIW, another way of conceptualizing this is the following:

While `apply` allows users and systems to cooperatively determine the desired state of an object, `kustomize` allows users and (other) humans to cooperatively determine the desired state of an object (using the same concepts and mechanism as apply does).

(For me) it was only after about a year of helping with efforts to fix bugs in apply and encountering issues that we had to fix as tacked-on imperative flags (e.g. `--prune`, `-n`) that we would need some declarative mechanism for these issues.  Repeatedly discussing the mental model of jointly owned resource / resource config between users and the system eventually sparked the notion of making empowering users with the same technique, but locally and declaratively.  (This was already sorta possible via `kubectl patch -f file.yaml -p '{"spec":{"some": "json"}}' --dry-run -o yaml`)

From this context we could frame 'Apply' as a mechanism to allow multiple parties to combine Resources + Resource Config together.  From this perspective, the kustomize functionality could be spiritually viewed as 'client-side apply' while the majority of what is today 'client-side apply' could be thought of as (what should should be) 'server-side apply'.

Reply:
On plugins: I don't think we've figured out what types of command-line-tool plugins are valuable and feasible. The technical difference between a standalone tool and executable plugin is small, and the challenges of building, testing, releasing, distributing, and updating tools/plugins for all of the OSes and environments that are supported by Kubernetes are similar in both cases.